### ISSUE TITLE
designdist faster than vegdist for binary distances 

### ISSUE THREAD

**Issue by gnutz** [vegandevs/vegan#182]
Hi vegan,

I've been calculating two binary distances (Bray & Jaccard) with the function `vegdist` on a reasonably big presence/absence matrix (2500 Columns). Recently, I have started implementing further distances with the function `designdist` and was surprised that this function was a lot faster than `vegdist`; even though there is a Note in the R Documentation of `designdist` that says:

> designdist does not use compiled code, and may be slow or use plenty of memory in large data sets.

 As the `vegdist` function often was some kind of bottleneck in my code, I was happy that I could speed it up this way. 
Here is a small benchmark of this behavior for the Bray distance; the same obviously holds for the Jaccard distance (and possibly other binary distances e.g. Ochiai...). 
```
library(rbenchmark)
library(vegan)

### Create (m,n)-matrix with random 0/1 values
m <- 1000
n <- m
set.seed(12345)
x <- sample.int (2, m*n, TRUE)-1L
dim(x) <- c(m,n)

### Calculate Bray Distance on x
vegdist.bray <- vegdist(x, method = "bray", binary = TRUE)
designdist.bray <- designdist(x, method = "(A+B-2*J)/(A+B)", 
                              terms = "binary")

### Are they equal?
all(vegdist.bray == designdist.bray)

### Benchmark it
bench.dist <- benchmark(vegdist.bray = vegdist(x, method = "bray", binary = TRUE),
                        designdist.bray = designdist(x, method = "(A+B-2*J)/(A+B)", terms = "binary"),
                        replications = 10, columns = c("test", "replications", "elapsed", "relative"), 
                        order = "elapsed")
print(bench.dist)
```
As this is not a bug, my only question would be if there is any drawback of using `designdist` instead of `vegdist`?

Kind regards

**Comment by jarioksa** [comment]
I have noticed the same, but I haven't run any tests. Vectorized **R** code is very fast at the present (it used to be slower). I cannot see any other disadvantage than the need of being very careful in writing dissimilarity formulae in `designdist`.

**Comment by jarioksa** [comment]
The difference seems to be much smaller and almost vanishing with quantitative indices. We use a very fast way of getting the terms with binary data (`tcrossprod` that is based on very fast compiled code), but for quantitative data with minimum terms we are much less optimal.

This means that something should be done to modernize `vegdist`.

**Comment by jarioksa** [comment]
I looked at the `vegdist()`  speed issues. It seems that currently `vegdist()` is 2 to 3 times slower than `dist()`. Currently `dist` is about as fast as `designdist()`. 

I tried to see how to speed up `vegdist()`, but DLL registration, byte compiling or parallel processing with `openMP` did not have much influence. I think that the main jump in the `dist` speed was gained in commit `r59032 | ripley | 2012-04-15 09:03:31 +0300 (Sun, 15 Apr 2012):  less copies in dist()` which adds a new wrapper function which copies only the really needed elements.

**Comment by jarioksa** [comment]
It seems now that a great part of relative slowness of `vegdist` is because `vegdist` checks input and can handle `NA` data, whereas `designdist`  fails  with `NA` (and this has been a concern in a [StackOverflow question](http://stackoverflow.com/questions/38930086/how-to-compute-ochiai-distance-matrix-with-pairwise-deletion-in-r)). If we only check for `NA` and `NaN` but allow `Inf` and `-Inf` then `vegdist` is about as fast as `stats::dist` and at least with `terms="minimum"` faster than `designdist`.

**Comment by jarioksa** [comment]
I was never so much concerned about `designdist` being fast, but I was worried about `vegdist` being so slow. I have now changed the text in `man/designdist.Rd` to tell that `designdist` indeed can be much faster like @gnutz suggested (commit 1b8532b449ff854089a704ba598790caf9076ccc). I have also made `vegdist` faster: it now matches `stats::dist` and is about two times faster than it used to be. However, it can still be much slower when `designdist` uses `"binary"` or `"quadratic"` terms. 

The following ideas could be explored:

* I could try to see if I can speed `terms="minimum"` in `designdist`. Preliminary tests indicate that a 20% speed-up is possible with dedicated compiled code. This would still make `terms = "minimum"` much slower than `terms="quadratic"` or `terms="binary"` in huge data sets. In smaller data sets of a few hundred observations the differences are small.

* We could consider adding a function of canned dissimilarity functions for a huge selection of possible indices. The authors would call functions by name, and the canned function would find the formula and call  `designdist`. It would be sufficient to provide a list of this style:
```R
canneddist <-
    function(x, method, terms)
{
index <- list(
    "sorensen" = list(method = "(A+B-2*J)/(A+B)", terms = "binary"),
    "bray" =   list(method = "(A+B-2*J)/(A+B)", terms = "minimum"),
    "whittaker" =  list(method = "(A+B-2*J)/(A+B)", terms = "binary"),
    "ochiai" = list(method = "1-J/sqrt(A*B)", terms = "binary"),
    "cosine" = list(method = "1-J/sqrt(A*B)", terms = "quadratic"))
ind <- match.arg(method, names(index))
if (is.na(ind))
    stop("unknown method")
z <- index[[ind]]
if (missing(terms)) # allow changing 'terms'
    terms <- z$terms
designdist(x, method = z$method, terms = terms, name = ind)
}
```

**Comment by jarioksa** [comment]
@gnutz : I have now made `terms = "minimum"` faster in `designdist()`. I wrote function finding the minimum terms in **C**, and now `designdist()` is *always* faster than `vegdist()`. I changed the documentation to say so without reservation. I close this issue, but I further the `canneddist` function by opening a new issue specifically on that (issue #196).

I don't think that we can make `vegdist` faster than `designdist` without radical rethinking. The vectorized code is so fast in **R** that simple compiled code cannot match that.
